t/proxyunits.t

```diff
@@ -69,9 +69,43 @@ for my $msrv (@mocksrvs) {
     push(@mbe, $be);
 }
 
-# Basic test with no backends. Write a command to the client, and check the
-# response.
+# $ps_send : request to proxy
+# $be_recv_ref : ref to a hashmap from be index to an array of received requests for validation.
+# $be_send_ref : ref to a hashmap from be index to an array of responses to proxy.
+# @ps_recv : response returned by proxy
+sub proxy_test {
```

Comment:
The reason why I don't do this more often is the test will fail at what's now an arbitrary line number, making it hard to find the actual failure point, and we can't make the messages as unique. The binary protocol tests are especially bad about this.

So if I want to abstract stuff it should return a pass/fail that can be checked with `ok(proxy_test(etc), "unique message")`

Reply:
there's also `subtest` for this type of pattern (see Test::More docs) but I don't use it often.

Reply:
That's a fair point. To improve, I passed the __LINE__ to `proxy_test`, so we won't lose insight.

Reply:
passing unique messages to `proxy_test` should also be easy to implement, but I don't see too much value at the moment.

Reply:
In dynamic languages, the line numbers can often be meaningless. That's why the perl test suite centers around the messages instead of line numbers; so please use that instead.

Reply:
Messages are also much more informative to end users than line numbers; if a user runs a test that fails and copy/pastes the information to me a line number means I also need to know the exact version/etc. It's much harder to deal with.

Reply:
> In dynamic languages, the line numbers can often be meaningless. 

Could you give me an example?

> so please use that instead.

Absolutely

Reply:
Fwiw the above example of calling proxy_test with `ok()` or using subtest are probably the cleanest; that lets proxy_test have unique messages that make sense internally, and a top level message that correlates with the overall test. 

Example: any code generation or function references aren't going to have line numbers that make sense. We don't typically use them in the test suite (look for `eval`), but centering around line numbers cuts us off from being able to use the language properly in these tests.

Also as I added earlier line numbers only make sense for the exact commit that generated them.

Reply:
Thanks for the explain. Subtest does seem to be the best option to me.

Migrated all to subtest, looks good.

```
# Subtest: request:ntokens()
    ok 1 - ps returned expected response.
    ok 2 - ps returned expected response.
    1..2
ok 112 - request:ntokens()
# Subtest: request:token() replacement
    ok 1 - be 0 received expected response
    ok 2 - be 0 received expected response
    ok 3 - ps returned expected response.
    1..3
```